File: dolby/src/co/aospa/dolby/xiaomi/DolbySettingsFragment.java
/*
 * Copyright (C) 2023-24 Paranoid Android
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package co.aospa.dolby.xiaomi;

import android.media.AudioAttributes;
import android.media.AudioDeviceAttributes;
import android.media.AudioDeviceCallback;
import android.media.AudioDeviceInfo;
import android.media.AudioManager;
import android.os.Bundle;
import android.os.Handler;
import android.util.Log;
import android.widget.CompoundButton;
import android.widget.Toast;

import androidx.preference.ListPreference;
import androidx.preference.Preference;
import androidx.preference.Preference.OnPreferenceChangeListener;
import androidx.preference.PreferenceFragment;
import androidx.preference.SwitchPreference;

import com.android.settingslib.widget.MainSwitchPreference;

import co.aospa.dolby.xiaomi.R;

import java.util.Arrays;
import java.util.List;

public class DolbySettingsFragment extends PreferenceFragment implements
        OnPreferenceChangeListener, CompoundButton.OnCheckedChangeListener {

    private static final String TAG = "DolbySettingsFragment";

    private static final AudioAttributes ATTRIBUTES_MEDIA = new AudioAttributes.Builder()
            .setUsage(AudioAttributes.USAGE_MEDIA)
            .build();

    public static final String PREF_ENABLE = "dolby_enable";
    public static final String PREF_PROFILE = "dolby_profile";
    public static final String PREF_PRESET = "dolby_preset";
    public static final String PREF_VIRTUALIZER = "dolby_virtualizer";
    public static final String PREF_STEREO = "dolby_stereo";
    public static final String PREF_DIALOGUE = "dolby_dialogue";
    public static final String PREF_BASS = "dolby_bass";
    public static final String PREF_VOLUME = "dolby_volume";
    public static final String PREF_RESET = "dolby_reset";

    private MainSwitchPreference mSwitchBar;
    private ListPreference mProfilePref, mPresetPref, mStereoPref, mDialoguePref;
    private SwitchPreference mBassPref, mVirtualizerPref, mVolumePref;
    private Preference mResetPref;
    private CharSequence[] mPresets, mDeValues, mSwValues;

    private DolbyUtils mDolbyUtils;
    private AudioManager mAudioManager;
    private boolean mDsOn, mIsOnSpeaker;
    private int mCurrentProfile = -1;
    private final Handler mHandler = new Handler();

    private final AudioDeviceCallback mAudioDeviceCallback = new AudioDeviceCallback() {
        public void onAudioDevicesAdded(AudioDeviceInfo[] addedDevices) {
            updateSpeakerState(false);
        }

        public void onAudioDevicesRemoved(AudioDeviceInfo[] removedDevices) {
            updateSpeakerState(false);
        }
    };

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        addPreferencesFromResource(R.xml.dolby_settings);

        mAudioManager = getActivity().getSystemService(AudioManager.class);
        mDolbyUtils = DolbyUtils.getInstance(getActivity());
        mDsOn = mDolbyUtils.getDsOn();

        mSwitchBar = (MainSwitchPreference) findPreference(PREF_ENABLE);
        mSwitchBar.addOnSwitchChangeListener(this);
        mSwitchBar.setChecked(mDsOn);

        mProfilePref = (ListPreference) findPreference(PREF_PROFILE);
        mProfilePref.setOnPreferenceChangeListener(this);
        mProfilePref.setEnabled(mDsOn);

        final CharSequence[] profiles = mProfilePref.getEntryValues();
        final int profile = mDolbyUtils.getProfile();
        if (Arrays.asList(profiles).contains(Integer.toString(profile))) {
            mCurrentProfile = profile;
            mProfilePref.setSummary("%s");
            mProfilePref.setValue(Integer.toString(profile));
        } else {
            mCurrentProfile = -1;
            mProfilePref.setSummary(getActivity().getString(R.string.dolby_unknown));
        }

        mPresetPref = (ListPreference) findPreference(PREF_PRESET);
        mPresetPref.setOnPreferenceChangeListener(this);
        mPresets = mPresetPref.getEntryValues();

        mVirtualizerPref = (SwitchPreference) findPreference(PREF_VIRTUALIZER);
        mVirtualizerPref.setOnPreferenceChangeListener(this);

        mStereoPref = (ListPreference) findPreference(PREF_STEREO);
        mStereoPref.setOnPreferenceChangeListener(this);
        mSwValues = mStereoPref.getEntryValues();

        mDialoguePref = (ListPreference) findPreference(PREF_DIALOGUE);
        mDialoguePref.setOnPreferenceChangeListener(this);
        mDeValues = mDialoguePref.getEntryValues();

        mBassPref = (SwitchPreference) findPreference(PREF_BASS);
        mBassPref.setOnPreferenceChangeListener(this);

        mVolumePref = (SwitchPreference) findPreference(PREF_VOLUME);
        mVolumePref.setOnPreferenceChangeListener(this);

        mResetPref = (Preference) findPreference(PREF_RESET);
        mResetPref.setOnPreferenceClickListener(p -> {
            mDolbyUtils.resetProfileSpecificSettings();
            updateProfileSpecificPrefs();
            Toast.makeText(getActivity(),
                    getActivity().getString(R.string.dolby_reset_profile_toast,
                            mProfilePref.getSummary()), Toast.LENGTH_SHORT).show();
            return true;
        });

        mAudioManager.registerAudioDeviceCallback(mAudioDeviceCallback, mHandler);
        updateSpeakerState(true);
    }

    @Override
    public void onDestroyView() {
        mAudioManager.unregisterAudioDeviceCallback(mAudioDeviceCallback);
        super.onDestroyView();
    }

    @Override
    public boolean onPreferenceChange(Preference preference, Object newValue) {
        switch (preference.getKey()) {
            case PREF_PROFILE:
                mCurrentProfile = Integer.parseInt((newValue.toString()));
                mDolbyUtils.setProfile(mCurrentProfile);
                updateProfileSpecificPrefs();
                return true;
            case PREF_PRESET:
                mDolbyUtils.setPreset(newValue.toString());
                return true;
            case PREF_VIRTUALIZER:
                if (mIsOnSpeaker)
                    mDolbyUtils.setSpeakerVirtualizerEnabled((Boolean) newValue);
                else
                    mDolbyUtils.setHeadphoneVirtualizerEnabled((Boolean) newValue);
                return true;
            case PREF_STEREO:
                mDolbyUtils.setStereoWideningAmount(Integer.parseInt((newValue.toString())));
                return true;
            case PREF_DIALOGUE:
                mDolbyUtils.setDialogueEnhancerAmount(Integer.parseInt((newValue.toString())));
                return true;
            case PREF_BASS:
                mDolbyUtils.setBassEnhancerEnabled((Boolean) newValue);
                return true;
            case PREF_VOLUME:
                mDolbyUtils.setVolumeLevelerEnabled((Boolean) newValue);
                return true;
            default:
                return false;
        }
    }

    @Override
    public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
        mDsOn = isChecked;
        mDolbyUtils.setDsOn(isChecked);
        mProfilePref.setEnabled(isChecked);
        mResetPref.setEnabled(isChecked);
        updateProfileSpecificPrefs();
    }

    private void updateSpeakerState(boolean force) {
        final AudioDeviceAttributes device =
                mAudioManager.getDevicesForAttributes(ATTRIBUTES_MEDIA).get(0);
        final boolean isOnSpeaker = (device.getType() == AudioDeviceInfo.TYPE_BUILTIN_SPEAKER);
        if (mIsOnSpeaker != isOnSpeaker || force) {
            dlog("updateSpeakerState: " + mIsOnSpeaker);
            mIsOnSpeaker = isOnSpeaker;
            updateProfileSpecificPrefs();
        }
    }

    private void updateProfileSpecificPrefs() {
        final String unknownRes = getActivity().getString(R.string.dolby_unknown);
        final String headphoneRes = getActivity().getString(R.string.dolby_connect_headphones);

        dlog("updateProfileSpecificPrefs: mDsOn=" + mDsOn
                + " mCurrentProfile=" + mCurrentProfile + " mIsOnSpeaker=" + mIsOnSpeaker);

        final boolean enable = mDsOn && (mCurrentProfile != -1);

        mPresetPref.setEnabled(enable);
        mVirtualizerPref.setEnabled(enable);
        mDialoguePref.setEnabled(enable);
        mVolumePref.setEnabled(enable);
        mResetPref.setEnabled(enable);

        mStereoPref.setEnabled(enable && !mIsOnSpeaker);
        mBassPref.setEnabled(enable && !mIsOnSpeaker);

        if (!enable) return;

        final String preset = mDolbyUtils.getPreset();
        if (Arrays.asList(mPresets).contains(preset)) {
            mPresetPref.setSummary("%s");
            mPresetPref.setValue(preset);
        } else {
            mPresetPref.setSummary(unknownRes);
        }

        final String deValue = Integer.toString(mDolbyUtils.getDialogueEnhancerAmount());
        if (Arrays.asList(mDeValues).contains(deValue)) {
            mDialoguePref.setSummary("%s");
            mDialoguePref.setValue(deValue);
        } else {
            mDialoguePref.setSummary(unknownRes);
        }

        mVirtualizerPref.setChecked(mIsOnSpeaker ? mDolbyUtils.getSpeakerVirtualizerEnabled()
                : mDolbyUtils.getHeadphoneVirtualizerEnabled());
        mVolumePref.setChecked(mDolbyUtils.getVolumeLevelerEnabled());

        if (mIsOnSpeaker) {
            mStereoPref.setSummary(headphoneRes);
            mBassPref.setSummary(headphoneRes);
            return;
        }

        final String swValue = Integer.toString(mDolbyUtils.getStereoWideningAmount());
        if (Arrays.asList(mSwValues).contains(swValue)) {
            mStereoPref.setSummary("%s");
            mStereoPref.setValue(swValue);
        } else {
            mStereoPref.setSummary(unknownRes);
        }

        mBassPref.setChecked(mDolbyUtils.getBassEnhancerEnabled());
        mBassPref.setSummary(null);
    }

    private static void dlog(String msg) {
        if (Log.isLoggable(TAG, Log.DEBUG)) {
            Log.d(TAG, msg);
        }
    }
}

File: dolby/src/co/aospa/dolby/xiaomi/DolbySettingsFragment.kt
/*
 * Copyright (C) 2023-24 Paranoid Android
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package co.aospa.dolby.xiaomi

import android.media.AudioAttributes
import android.media.AudioDeviceCallback
import android.media.AudioDeviceInfo
import android.media.AudioManager
import android.os.Bundle
import android.os.Handler
import android.widget.CompoundButton
import android.widget.CompoundButton.OnCheckedChangeListener
import android.widget.Toast
import androidx.preference.ListPreference
import androidx.preference.Preference
import androidx.preference.Preference.OnPreferenceChangeListener
import androidx.preference.PreferenceFragment
import androidx.preference.SwitchPreferenceCompat
import co.aospa.dolby.xiaomi.DolbyConstants.Companion.dlog
import co.aospa.dolby.xiaomi.R
import com.android.settingslib.widget.MainSwitchPreference

class DolbySettingsFragment : PreferenceFragment(),
    OnPreferenceChangeListener, CompoundButton.OnCheckedChangeListener {

    private val switchBar by lazy {
        findPreference<MainSwitchPreference>(DolbyConstants.PREF_ENABLE)!!
    }
    private val profilePref by lazy {
        findPreference<ListPreference>(DolbyConstants.PREF_PROFILE)!!
    }
    private val presetPref by lazy {
        findPreference<ListPreference>(DolbyConstants.PREF_PRESET)!!
    }
    private val stereoPref by lazy {
        findPreference<ListPreference>(DolbyConstants.PREF_STEREO)!!
    }
    private val dialoguePref by lazy {
        findPreference<ListPreference>(DolbyConstants.PREF_DIALOGUE)!!
    }
    private val bassPref by lazy {
        findPreference<SwitchPreferenceCompat>(DolbyConstants.PREF_BASS)!!
    }
    private val virtPref by lazy {
        findPreference<SwitchPreferenceCompat>(DolbyConstants.PREF_VIRTUALIZER)!!
    }
    private val volumePref by lazy {
        findPreference<SwitchPreferenceCompat>(DolbyConstants.PREF_VOLUME)!!
    }
    private val resetPref by lazy {
        findPreference<Preference>(DolbyConstants.PREF_RESET)!!
    }

    private val dolbyController by lazy { DolbyController.getInstance(context) }
    private val audioManager by lazy { context.getSystemService(AudioManager::class.java) }
    private val handler = Handler()

    private var isOnSpeaker = true
        set(value) {
            if (field == value) return
            field = value
            dlog(TAG, "setIsOnSpeaker($value)")
            updateProfileSpecificPrefs()
        }

    private val audioDeviceCallback = object : AudioDeviceCallback() {
        override fun onAudioDevicesAdded(addedDevices: Array<AudioDeviceInfo>) {
            dlog(TAG, "onAudioDevicesAdded")
            updateSpeakerState()
        }

        override fun onAudioDevicesRemoved(removedDevices: Array<AudioDeviceInfo>) {
            dlog(TAG, "onAudioDevicesRemoved")
            updateSpeakerState()
        }
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        dlog(TAG, "onCreatePreferences")
        addPreferencesFromResource(R.xml.dolby_settings)

        val dsOn = dolbyController.dsOn
        switchBar.addOnSwitchChangeListener(this)
        switchBar.setChecked(dsOn)

        profilePref.onPreferenceChangeListener = this
        profilePref.setEnabled(dsOn)

        val profile = dolbyController.profile
        profilePref.apply {
            if (entryValues.contains(profile.toString())) {
                summary = "%s"
                value = profile.toString()
            } else {
                summary = context.getString(R.string.dolby_unknown)
            }
        }

        presetPref.onPreferenceChangeListener = this
        virtPref.onPreferenceChangeListener = this
        stereoPref.onPreferenceChangeListener = this
        dialoguePref.onPreferenceChangeListener = this
        bassPref.onPreferenceChangeListener = this
        volumePref.onPreferenceChangeListener = this

        resetPref.setOnPreferenceClickListener {
            dolbyController.resetProfileSpecificSettings()
            updateProfileSpecificPrefs()
            Toast.makeText(
                context,
                context.getString(R.string.dolby_reset_profile_toast, profilePref.summary),
                Toast.LENGTH_SHORT
            ).show()
            true
        }

        audioManager.registerAudioDeviceCallback(audioDeviceCallback, handler)
        updateSpeakerState()
        updateProfileSpecificPrefs()
    }

    override fun onDestroyView() {
        dlog(TAG, "onDestroyView")
        audioManager.unregisterAudioDeviceCallback(audioDeviceCallback)
        super.onDestroyView()
    }

    override fun onPreferenceChange(preference: Preference, newValue: Any): Boolean {
        dlog(TAG, "onPreferenceChange: key=${preference.key} value=$newValue")
        when (preference.key) {
            DolbyConstants.PREF_PROFILE -> {
                dolbyController.profile = newValue.toString().toInt()
                updateProfileSpecificPrefs()
            }

            DolbyConstants.PREF_PRESET -> {
                dolbyController.preset = newValue.toString()
            }

            DolbyConstants.PREF_VIRTUALIZER -> {
                if (isOnSpeaker)
                    dolbyController.speakerVirtEnabled = newValue as Boolean
                else
                    dolbyController.headphoneVirtEnabled = newValue as Boolean
            }

            DolbyConstants.PREF_STEREO -> {
                dolbyController.stereoWideningAmount = newValue.toString().toInt()
            }

            DolbyConstants.PREF_DIALOGUE -> {
                dolbyController.dialogueEnhancerAmount = newValue.toString().toInt()
            }

            DolbyConstants.PREF_BASS -> {
                dolbyController.bassEnhancerEnabled = newValue as Boolean
            }

            DolbyConstants.PREF_VOLUME -> {
                dolbyController.volumeLevelerEnabled = newValue as Boolean
            }

            else -> return false
        }
        return true
    }

    override fun onCheckedChanged(buttonView: CompoundButton, isChecked: Boolean) {
        dlog(TAG, "onCheckedChanged($isChecked)")
        dolbyController.dsOn = isChecked
        profilePref.setEnabled(isChecked)
        updateProfileSpecificPrefs()
    }

    private fun updateSpeakerState() {
        val device = audioManager.getDevicesForAttributes(ATTRIBUTES_MEDIA)[0]
        isOnSpeaker = (device.type == AudioDeviceInfo.TYPE_BUILTIN_SPEAKER)
    }

    private fun updateProfileSpecificPrefs() {
        val unknownRes = context.getString(R.string.dolby_unknown)
        val headphoneRes = context.getString(R.string.dolby_connect_headphones)
        val dsOn = dolbyController.dsOn
        val currentProfile = dolbyController.profile

        dlog(
            TAG, "updateProfileSpecificPrefs: dsOn=$dsOn currentProfile=$currentProfile"
                    + " isOnSpeaker=$isOnSpeaker"
        )

        val enable = dsOn && (currentProfile != -1)
        presetPref.setEnabled(enable)
        virtPref.setEnabled(enable)
        dialoguePref.setEnabled(enable)
        volumePref.setEnabled(enable)
        resetPref.setEnabled(enable)
        stereoPref.setEnabled(enable && !isOnSpeaker)
        bassPref.setEnabled(enable && !isOnSpeaker)

        if (!enable) return

        val preset = dolbyController.preset
        presetPref.apply {
            if (entryValues.contains(preset)) {
                summary = "%s"
                value = preset
            } else {
                summary = unknownRes
            }
        }

        val deValue = dolbyController.dialogueEnhancerAmount.toString()
        dialoguePref.apply {
            if (entryValues.contains(deValue)) {
                summary = "%s"
                value = deValue
            } else {
                summary = unknownRes
            }
        }

        virtPref.setChecked(if (isOnSpeaker) {
            dolbyController.speakerVirtEnabled
        } else {
            dolbyController.headphoneVirtEnabled
        })

        volumePref.setChecked(dolbyController.volumeLevelerEnabled)

        // below prefs are not enabled on loudspeaker
        if (isOnSpeaker) {
            stereoPref.summary = headphoneRes
            bassPref.summary = headphoneRes
            return
        }

        val swValue = dolbyController.stereoWideningAmount.toString()
        stereoPref.apply {
            if (entryValues.contains(swValue)) {
                summary = "%s"
                value = swValue
            } else {
                summary = unknownRes
            }
        }

        bassPref.apply {
            setChecked(dolbyController.bassEnhancerEnabled)
            summary = null
        }
    }

    companion object {
        private const val TAG = "DolbySettingsFragment"
        private val ATTRIBUTES_MEDIA = AudioAttributes.Builder()
                .setUsage(AudioAttributes.USAGE_MEDIA)
                .build()
    }
}
